File: frontend/src/types/flag.ts
import { z } from "zod"
import { StudentSchema } from "./student"
import { UserSchema } from "./user"

export const FlagTypeSchema = z.enum([
	"REPORTING",
	"VALIDATION"
])
export type FlagType = z.infer<typeof FlagTypeSchema>

export const FlagSchema = z.object({
	id: z.number(),
	description: z.string().nullable(),
	type: FlagTypeSchema,
	firstStudent: StudentSchema.nullable(),
	secondStudent: StudentSchema.nullable(),
	author: UserSchema,
	status: z.boolean().nullable()
})
export type Flag = z.infer<typeof FlagSchema>

export const CreateFlagSchema = FlagSchema.omit({
	id: true,
	firstStudent: true,
	secondStudent: true,
	author: true,
	project: true,
	status: true
}).extend({
	firstStudentId: z.coerce.number().optional(),
	secondStudentId: z.coerce.number().optional(),
	authorId: z.coerce.number(),
	projectId: z.coerce.number(),
	description: z.string().optional()
})
export type CreateFlag = z.infer<typeof CreateFlagSchema>

export const UpdateFlagSchema = CreateFlagSchema.partial()
export type UpdateFlag = z.infer<typeof UpdateFlagSchema>

File: backend/src/main/java/fr/eseo/tauri/model/Flag.java
package fr.eseo.tauri.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import fr.eseo.tauri.model.enumeration.FlagType;
import fr.eseo.tauri.util.valid.Create;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Data;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import java.io.Serializable;

@Entity
@Table(name = "flags")
@Data
public class Flag implements Serializable {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @JsonProperty
    private Integer id;

    @JsonProperty
    private String description;

    @NotNull(groups = { Create.class }, message = "The type field is required")
    @Enumerated(EnumType.STRING)
    @Column(name="type")
    @JsonProperty
    private FlagType type;

    @JsonProperty
    private Boolean status;

    @ManyToOne
    @JoinColumn(name = "first_student_id")
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JsonProperty
    private Student firstStudent;

    @ManyToOne
    @JoinColumn(name = "second_student_id")
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JsonProperty
    private Student secondStudent;

    @ManyToOne
    @JoinColumn(name = "author_id")
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JsonProperty
    private User author;

    @ManyToOne
    @JoinColumn(name = "project_id")
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JsonProperty
    private Project project;

    @Transient
    @JsonDeserialize
    private Integer firstStudentId;

    @Transient
    @JsonDeserialize
    private Integer secondStudentId;

    @NotNull(groups = { Create.class }, message = "The authorId field is required")
    @Transient
    @JsonDeserialize
    private Integer authorId;

    @NotNull(groups = { Create.class }, message = "The projectId field is required")
    @Transient
    @JsonDeserialize
    private Integer projectId;

}
